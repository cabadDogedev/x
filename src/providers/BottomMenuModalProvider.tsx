import React, { createContext, useMemo } from 'react';

// components
import { SendModalData } from '../components/BottomMenuModal/SendModal';

export interface BottomMenuModalContext {
  data: {
    showTransactionConfirmation: (payload: SendModalData) => void;
    showSend: () => void;
    showHistory: () => void;
    showAccount:() => void;
    showApps: () => void;
    hide: () => void;
    active: BottomMenuItem | null;
  }
}

export const ProviderContext = createContext<BottomMenuModalContext | null>(null);

interface BottomMenuSend {
  type: 'send';
  payload?: SendModalData
}

export type BottomMenuItem = {
  type: 'history' | 'account' | 'apps';
} | BottomMenuSend;

const BottomMenuModalProvider = ({ children }: React.PropsWithChildren) => {
  const [activeMenuItem, setActiveMenuItem] = React.useState<BottomMenuItem | null>(null);

  const hide = () => setActiveMenuItem(null);

  const contextData = useMemo(() => ({
    showTransactionConfirmation: (payload?: SendModalData) => setActiveMenuItem({ type: 'send', payload }),
    showSend: () => setActiveMenuItem({ type: 'send' }),
    showHistory: () => setActiveMenuItem({ type: 'history' }),
    showAccount: () => setActiveMenuItem({ type: 'account' }),
    showApps: () => setActiveMenuItem({ type: 'apps' }),
    hide,
    active: activeMenuItem
  }), [activeMenuItem]);

  return (
    <ProviderContext.Provider value={{ data: contextData }}>
      {children}
    </ProviderContext.Provider>
  );
}

export default BottomMenuModalProvider;
