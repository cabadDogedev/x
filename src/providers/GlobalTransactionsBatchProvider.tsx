import React, { createContext, useMemo } from 'react';

// utils
import { getObjectHash } from '../utils/common';

// types
import { ITransaction } from '../types/blockchain';

export type IGlobalBatchTransaction = {
  title: string;
  description?: string;
  id?: string
} & ITransaction;

export interface IGlobalTransactionsBatchContext {
  data: {
    transactions: IGlobalBatchTransaction[];
    addToBatch: (transaction: IGlobalBatchTransaction) => void;
    removeFromBatch: (transactionId: string) => void;
  }
}

export const GlobalTransactionsBatchContext = createContext<IGlobalTransactionsBatchContext | null>(null);

const GlobalTransactionsBatchProvider = ({ children }: { children: React.ReactNode }) => {
  const [transactions, setBatches] = React.useState<IGlobalBatchTransaction[]>([]);

  const addToBatch = (transaction: IGlobalBatchTransaction) => {
    setBatches((prev) => prev.concat({
      ...transaction,
      id: transaction.id || getObjectHash(transaction, +new Date() + Math.random()),
    }));
  }

  const removeFromBatch = (transactionId: string) => {
    setBatches((prev) => prev.filter((tx) => tx.id !== transactionId));
  }

  const contextData = useMemo(() => ({
    transactions,
    addToBatch,
    removeFromBatch,
  }), [
    transactions,
  ]);

  return (
    <GlobalTransactionsBatchContext.Provider value={{ data: contextData }}>
      {children}
    </GlobalTransactionsBatchContext.Provider>
  );
}

export default GlobalTransactionsBatchProvider;
