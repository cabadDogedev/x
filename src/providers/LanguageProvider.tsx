import i18n from 'i18next';
import React from 'react';
import { I18nextProvider, initReactI18next } from 'react-i18next';

// translations
import translationEn from '../translations/en.json';

i18n.use(initReactI18next).init({
  lng: 'en',
  fallbackLng: 'en',
  resources: {
    en: { translation: translationEn },
  },
  interpolation: {
    escapeValue: false, // https://www.i18next.com/translation-function/interpolation#unescape
  },
});

const LanguageProvider = ({ children }: { children: React.ReactNode }) => {
  return <I18nextProvider i18n={i18n}>{children}</I18nextProvider>;
};

export default LanguageProvider;
