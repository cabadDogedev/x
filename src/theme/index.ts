import { createGlobalStyle, keyframes } from 'styled-components';

// fonts
import formularRegularFont from '../assets/fonts/Formular-Regular.otf';
import formularMediumFont from '../assets/fonts/Formular-Medium.otf';
import formularBoldFont from '../assets/fonts/Formular-Bold.otf';

export interface Theme {
  font: {
    primary: string;
  }
  color: {
    background: {
      body: string;
      bottomMenu: string;
      bottomMenuItemHover: string;
      bottomMenuModal: string;
      loadingLogo: string;
      buttonPrimary: string;
      buttonPrimaryDisabled: string;
      input: string;
      info: string;
      error: string;
      success: string;
      warning: string;
      horizontalDivider: string;
      selectItem: string;
      selectItemHover: string;
      bottomMenuItemNotification: string;
    },
    text: {
      body: string;
      bottomMenuItem: string;
      loadingLogo: string;
      buttonPrimary: string;
      buttonPrimaryDisabled: string;
      input: string;
      info: string;
      error: string;
      success: string;
      warning: string;
      inputPlaceholder: string;
      inputHelper: string;
    },
    border: {
      bottomMenuItemBottomActive: string;
    },
    icon: {
      inputHelper: string;
      delete: string;
    }
  }
}

export const defaultTheme: Theme = {
  font: {
    primary: 'Formular, serif',
  },
  color: {
    background: {
      body: '#101010',
      bottomMenu: 'rgba(16,16,16,0.7)',
      bottomMenuItemHover: 'rgba(216, 232, 255, 0.10)',
      bottomMenuModal: 'rgba(16, 16, 16, 0.70)',
      loadingLogo: '#fff',
      buttonPrimary: '#D9D9D9',
      buttonPrimaryDisabled: '#63626B',
      input: 'rgba(217, 217, 217, 0.05)',
      info: '#BEF',
      error: '#D8000C',
      success: '#DFF2BF',
      warning: '#FEEFB3',
      horizontalDivider: 'linear-gradient(270deg, rgba(217, 217, 217, 0.00) 0%, #D9D9D9 50.52%, rgba(217, 217, 217, 0.00) 100%)',
      selectItem: 'rgba(217, 217, 217, 0.05)',
      selectItemHover: 'rgba(217, 217, 217, 0.2)',
      bottomMenuItemNotification: 'rgb(0,111,255)',
    },
    text: {
      body: '#fff',
      bottomMenuItem: '#fff',
      loadingLogo: '#997cfa',
      buttonPrimary: '#1D1D1D',
      buttonPrimaryDisabled: '#1D1D1D',
      input: '#fff',
      info: '#059',
      error: '#FFBABA',
      success: '#270',
      warning: '#9F6000',
      inputPlaceholder: '#838383',
      inputHelper: 'rgba(255, 255, 255, 0.3)',
    },
    border: {
      bottomMenuItemBottomActive: '#fff',
    },
    icon: {
      inputHelper: '#fff',
      delete: '#ff000080',
    }
  },
};

const skeleton = keyframes`
  0% {
    background-color: hsl(0, 0%, 27%);
  }
  100% {
    background-color: hsl(0, 0%, 53%);
  }
`;


const pulse = (min = 1, max = 2) => keyframes`
  0% {
    transform: scale(${min});
  }
  50% {
    transform: scale(${max});
  }
  100% {
    transform: scale(${min});
  }
`;

const rotateAndPulse = keyframes`
  0% {
    transform: scale(1) rotate(0deg);
  }
  50% {
    transform: scale(1.5) rotate(180deg);
  }
  100% {
    transform: scale(1) rotate(360deg);
  }
`;

const slideUp = keyframes`
  0% {
    transform: scale(.8) translateY(1000px);
    opacity: 0;
  }
  100% {
    transform: scale(1) translateY(0px);
    opacity: 1;
  }
`;

const slideDown = keyframes`
  0% {
    transform: scale(1) translateY(0px);
    opacity: 1;
  }
  100% {
    transform: scale(.8) translateY(1000px);
    opacity: 0;
  }
`;

export const animation = {
  skeleton,
  pulse,
  rotateAndPulse,
  slideUp,
  slideDown,
}

export const GlobalStyle = createGlobalStyle`
  @font-face {
    font-family: Formular;
    font-weight: 400;
    src: url(${formularRegularFont}) format("opentype");
  }

  @font-face {
    font-family: Formular;
    font-weight: 500;
    src: url(${formularMediumFont}) format("opentype");
  }

  @font-face {
    font-family: Formular;
    font-weight: 700;
    src: url(${formularBoldFont}) format("opentype");
  }

  body {
    font-family: ${({ theme }) => theme.font.primary};
    background: ${({ theme }) => theme.color.background.body};
    color: ${({ theme }) => theme.color.text.body};
  }
  
  input, textarea, button, select {
    font-family: ${({ theme }) => theme.font.primary};
  }
  
  * {
    box-sizing: border-box;
    line-height: normal;
    margin: 0;
    padding: 0;
  }
`;

