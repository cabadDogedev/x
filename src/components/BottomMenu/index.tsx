import { usePrivy } from '@privy-io/react-auth';
import { useTranslation } from 'react-i18next';
import { useLocation, useNavigate } from 'react-router-dom';
import styled from 'styled-components';

// images
import { ReactComponent as IconApps } from '../../assets/images/icon/icon-apps.svg';
import { ReactComponent as IconHistory } from '../../assets/images/icon/icon-history.svg';
import { ReactComponent as IconHome } from '../../assets/images/icon/icon-home.svg';
import { ReactComponent as IconSend } from '../../assets/images/icon/icon-send.svg';
import { ReactComponent as IconWallet } from '../../assets/images/icon/icon-wallet.svg';

// navigation
import { navigationRoute } from '../../navigation';

// hooks
import useBottomMenuModal from '../../hooks/useBottomMenuModal';
import useGlobalTransactionsBatch from '../../hooks/useGlobalTransactionsBatch';

// components
import BottomMenuModal from '../BottomMenuModal';

// theme
import { animation } from '../../theme';

const BottomMenu = () => {
  const { authenticated } = usePrivy();
  const navLocation = useLocation();
  const navigate = useNavigate();
  const [t] = useTranslation();
  const { active, showSend, showApps, showHistory, showAccount, hide } = useBottomMenuModal();
  const { transactions: globalTransactionsBatch } = useGlobalTransactionsBatch();

  if (!authenticated) return null;

  const menuItems = [
    {
      icon: <IconSend />,
      type: 'send',
      iconNotificationCounter: globalTransactionsBatch.length,
      label: t`menuAction.send`,
      show: showSend,
    },
    {
      icon: <IconHistory />,
      type: 'history',
      label: t`menuAction.history`,
      show: showHistory,
    },
    {
      icon: <IconWallet />,
      type: 'account',
      label: t`menuAction.account`,
      show: showAccount,
    },
    {
      icon: <IconApps />,
      type: 'apps',
      label: t`menuAction.apps`,
      show: showApps,
    },
  ];

  const isHomeActive = active === null
    && navLocation.pathname === '/';

  return (
    <>
      <Wrapper>
        <HomeMenuItem>
          <MenuItem
            onClick={() => navigate(navigationRoute.home)}
            className={isHomeActive ? 'active' : ''}
          >
            <IconHome />
          </MenuItem>
        </HomeMenuItem>
        <MainMenuItems>
          {menuItems.map((item, index) => (
            <MenuItem
              key={item.label + index}
              onClick={() => {
                if (active?.type === item.type) {
                  hide();
                  return;
                }
                item.show();
              }}
              className={active?.type === item.type ? 'active' : ''}
            >
              {item.icon}
              {!!item.iconNotificationCounter && (
                <MenuItemNotification>{item.iconNotificationCounter}</MenuItemNotification>
              )}
              <span>{item.label}</span>
            </MenuItem>
          ))}
        </MainMenuItems>
      </Wrapper>
      <BottomMenuModal
        activeMenuItem={active}
        onClose={hide}
      />
    </>
  );
}

const MenuItemNotification = styled.div`
  position: absolute;
  transition: all .1s ease-in-out;
  bottom: -4px;
  left: 15px;
  background: ${({ theme }) => theme.color.background.bottomMenuItemNotification};
  padding: 2px;
  text-align: center;
  width: 17px;
  height: 17px;
  border-radius: 50%;
  color: #fff;
  font-size: 10px;
  box-shadow: 0 0 0 0 rgba(0, 0, 0, 1);
  transform: scale(1);
  animation: ${animation.pulse(0.85, 1)} 2s infinite;
`;

const MenuItem = styled.div`
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 12px;
  color: ${({ theme }) => theme.color.text.bottomMenuItem};
  height: 48px;
  cursor: pointer;
  transition: all .1s ease-in-out;
  font-weight: 400;
  letter-spacing: -0.5px;
  font-size: 14px;

  &:first-child {
    margin-left: 17px;
  }

  &:last-child {
    margin-right: 17px;
  }

  span {
    display: none;
  }
`;

const Wrapper = styled.div`
  margin: 0 auto;
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 10px;
  position: fixed;
  bottom: 40px;
  left: 50%;
  z-index: 100;
  transform: translateX(-50%);

  &:not(&:hover) ${MenuItem}.active, ${MenuItem}:hover {
    padding: 0 13px;
    margin: 0;
    border-radius: 100px;
    background: ${({ theme }) => theme.color.background.bottomMenuItemHover};
    
    ${MenuItemNotification} {
      left: 28px;
    }

    span {
      display: block;
    }
  }
`;

const MainMenuItems = styled.div`
  flex: 1;
  display: flex;
  align-items: center;
  justify-content: space-between;
  backdrop-filter: blur(5px);
  background: ${({ theme }) => theme.color.background.bottomMenu};
  border-radius: 130px;
  padding: 6px 7px;
  width: 280px;
  height: 60px;
`;

const HomeMenuItem = styled(MainMenuItems)`
  flex: none;
  width: 60px;
  height: 60px;

  ${MenuItem} {
    width: 48px;

    &:first-child, &:last-child {
      margin: 0
    }

    &:hover {
      padding: 0;
    }
  }
`;

export default BottomMenu;
