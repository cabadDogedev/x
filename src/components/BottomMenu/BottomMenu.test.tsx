import renderer, { ReactTestRendererJSON, ReactTestRenderer, act } from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import { BrowserRouter } from 'react-router-dom';
import { usePrivy } from '@privy-io/react-auth';
import { EtherspotTransactionKit } from '@etherspot/transaction-kit';
import { ethers } from 'ethers';

// components
import BottomMenu from './';

// theme
import { defaultTheme } from '../../theme';

// providers
import BottomMenuModalProvider from '../../providers/BottomMenuModalProvider';
import LanguageProvider from '../../providers/LanguageProvider';
import GlobalTransactionsBatchProvider from '../../providers/GlobalTransactionsBatchProvider';

const ethersProvider = new ethers.providers.JsonRpcProvider('http://localhost:8545', 'goerli'); // replace with your node's RPC URL
const provider = ethers.Wallet.createRandom().connect(ethersProvider);

describe('<BottomMenu />', () => {
  let rendered: ReactTestRenderer | null = null;

  it('renders correctly', async () => {
    (usePrivy as jest.Mock).mockImplementation(() => ({ authenticated: true }));

    await act(async () => {
      rendered = renderer
        .create(
          <BrowserRouter>
            <EtherspotTransactionKit provider={provider}>
              <ThemeProvider theme={defaultTheme}>
                <LanguageProvider>
                  <GlobalTransactionsBatchProvider>
                    <BottomMenuModalProvider>
                      <BottomMenu />
                    </BottomMenuModalProvider>
                  </GlobalTransactionsBatchProvider>
                </LanguageProvider>
              </ThemeProvider>
            </EtherspotTransactionKit>
          </BrowserRouter>
        )
    });

    const tree = rendered?.toJSON();
    expect(tree).toMatchSnapshot();

    const [bottomMenuElement] = tree as ReactTestRendererJSON[];

    expect((bottomMenuElement.children?.[1] as ReactTestRendererJSON)?.children?.length).toBe(5); // other menu items
    expect(bottomMenuElement.type).toBe('div');

    jest.clearAllMocks();
  });

  it('renders correctly if not authenticated', async () => {
    (usePrivy as jest.Mock).mockImplementation(() => ({ authenticated: false }));

    await act(async () => {
      rendered = renderer
        .create(
          <BrowserRouter>
            <ThemeProvider theme={defaultTheme}>
              <LanguageProvider>
                <GlobalTransactionsBatchProvider>
                  <BottomMenuModalProvider>
                    <BottomMenu />
                  </BottomMenuModalProvider>
                </GlobalTransactionsBatchProvider>
              </LanguageProvider>
            </ThemeProvider>
          </BrowserRouter>
        )
    });

    const tree = rendered?.toJSON();
    expect(tree).toMatchSnapshot();

    const treeElement = tree as ReactTestRendererJSON;
    expect(treeElement).toBe(null);

    jest.clearAllMocks();
  });
});


