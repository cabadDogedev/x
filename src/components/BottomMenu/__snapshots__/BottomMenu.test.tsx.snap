// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BottomMenu /> renders correctly 1`] = `
[
  .c4 {
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 12px;
  color: #fff;
  height: 48px;
  cursor: pointer;
  transition: all .1s ease-in-out;
  font-weight: 400;
  letter-spacing: -0.5px;
  font-size: 14px;
}

.c4:first-child {
  margin-left: 17px;
}

.c4:last-child {
  margin-right: 17px;
}

.c4 span {
  display: none;
}

.c0 {
  margin: 0 auto;
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 10px;
  position: fixed;
  bottom: 40px;
  left: 50%;
  z-index: 100;
  transform: translateX(-50%);
}

.c0:not(.c0:hover) .c3.active,
.c0 .c3:hover {
  padding: 0 13px;
  margin: 0;
  border-radius: 100px;
  background: rgba(216, 232, 255, 0.10);
}

.c0:not(.c0:hover) .c3.active span,
.c0 .c3:hover span {
  display: block;
}

.c1 {
  flex: 1;
  display: flex;
  align-items: center;
  justify-content: space-between;
  backdrop-filter: blur(5px);
  background: rgba(16,16,16,0.7);
  border-radius: 130px;
  padding: 6px 7px;
  width: 280px;
  height: 60px;
}

.c2 {
  flex: none;
  width: 60px;
  height: 60px;
}

.c2 .c3 {
  width: 48px;
}

.c2 .c3:first-child,
.c2 .c3:last-child {
  margin: 0;
}

.c2 .c3:hover {
  padding: 0;
}

<div
    className="c0"
  >
    <div
      className="c1 c2"
    >
      <div
        className="c3 c4 active"
        onClick={[Function]}
      >
        <svg />
      </div>
    </div>
    <div
      className="c1"
    >
      <div
        className="c3 c4"
        onClick={[Function]}
      >
        <svg />
        <span>
          Send
        </span>
      </div>
      <div
        className="c3 c4"
        onClick={[Function]}
      >
        <svg />
        <span>
          History
        </span>
      </div>
      <div
        className="c3 c4"
        onClick={[Function]}
      >
        <svg />
        <span>
          Account
        </span>
      </div>
      <div
        className="c3 c4"
        onClick={[Function]}
      >
        <svg />
        <span>
          Apps
        </span>
      </div>
    </div>
  </div>,
  .c0 {
  position: fixed;
  z-index: 98;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  transition: 100ms linear;
  backdrop-filter: blur(0px);
  display: none;
  align-items: flex-end;
  justify-content: center;
  padding-bottom: 114px;
  overflow: hidden;
}

.c1 {
  transition: 100ms linear;
  transform: translateY(1000px);
  display: flex;
  flex-direction: row;
  align-content: start;
  justify-content: start;
  width: 100%;
}

.c2 {
  align-self: flex-start;
  transform: translateX(calc(0 * -100vw));
  display: flex;
  flex-direction: row;
  align-content: start;
  justify-content: start;
}

.c3 {
  width: 100vw;
  display: flex;
  align-content: center;
  justify-content: center;
}

.c4 {
  width: 350px;
  min-height: 350px;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  border-radius: 32px;
  padding: 34px 23px;
  background: rgba(16, 16, 16, 0.70);
  backdrop-filter: blur(5px);
  overflow: hidden;
}

<div
    className="c0"
  >
    <div
      className="c1"
    >
      <div
        className="c2"
      >
        <div
          className="c3"
        >
          <div
            className="c4"
          />
        </div>
        <div
          className="c3"
        >
          <div
            className="c4"
          />
        </div>
        <div
          className="c3"
        >
          <div
            className="c4"
          />
        </div>
        <div
          className="c3"
        >
          <div
            className="c4"
          />
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`<BottomMenu /> renders correctly if not authenticated 1`] = `null`;
